// regexp-support/test/bin/cache.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const path = require("path");
const globby = require("globby");
const fs = require("fs-extra");
exports.REGEXPP_PATH = path.dirname(require.resolve('regexpp2/package.json'));
exports.REGEXPP_PATH_DATA = path.join(exports.REGEXPP_PATH, 'src/unicode/property-data');
exports.XREGEXP_PATH = path.dirname(require.resolve('xregexp/package.json'));
exports.XREGEXP_PATH_DATA = path.join(exports.XREGEXP_PATH, 'tools/output');
(async () => {
    const PropertyData = await Promise.resolve().then(() => require(exports.REGEXPP_PATH_DATA)).then(v => v.default);
    let output = [];
    let def = [
        '$LONE',
        'General_Category',
        'Script',
    ];
    def.forEach(function (key) {
        //console.log(PropertyData[key]);
        let d = [...PropertyData[key]]
            .reduce(function (a, b) {
            a[b] = false;
            return a;
        }, {});
        let s = JSON.stringify(d, null, "\t");
        output.push(`export const ${key} = ${s};`);
    });
    output.push(`export default { ${def.join(', ')} }`);
    let name = 'property-data';
    let file = path.join(__dirname, '../../', 'lib/pattern/cache', name + '.ts');
    fs.outputFile(file, output.join('\n\n'));
})();
globby([
    '*.js',
], {
    cwd: exports.XREGEXP_PATH_DATA,
})
    .then((ls) => {
    ls.forEach(async function (value) {
        let name = path.parse(value).name;
        let data = await getData(name);
        let aa = {};
        let d = Object
            .values(data)
            .reduce(function (a, b) {
            a[b.name] = false;
            if (b.alias) {
                a[b.alias] = a[b.name];
                aa[b.name] = b.alias;
                aa[b.alias] = b.name;
            }
            return a;
        }, {});
        let s = JSON.stringify(d, null, "\t");
        fs.outputFile(path.join(__dirname, '../../', 'lib/pattern/cache', name + '.ts'), `export default ${s};\n\nexport const NAME_ALIAS = ${JSON.stringify(aa, null, "\t")};`);
    });
});
function getData(name) {
    let file = path.join(exports.XREGEXP_PATH_DATA, name);
    return Promise.resolve().then(() => require(file));
}
exports.getData = getData;
